src/adaptation.jl

```diff
@@ -0,0 +1,231 @@
+# -------------
+
+# Non-linear Bradford
+struct CAT_BFD_NL <: NonLinearCAT
+end
+
+
+
+# Define "domain-specific" conversion methods to simplify the conversion chains
+
+# xyz_to_lms()
+# ------------
+function xyz_to_lms(c::XYZ{T}, ::XYZ, cat::MatrixCAT) where T
+    @mul3x3 LMS{T} cat_cnvt(cat) c.x c.y c.z
+end
+
+function xyz_to_lms(c::XYZ{T}, wp::XYZ, cat::CAT_BFD_NL) where T
+    wp_lms = @mul3x3xyz LMS{T} BFD wp
+    li_lms = @mul3x3xyz LMS{T} BFD c
+    p = Float64(wp_lms.s) ^ 0.0834
+    LMS{T}( li_lms.l / wp_lms.l,
+            li_lms.m / wp_lms.m,
+           (li_lms.s / wp_lms.s)^p )
+end
+
+
+# lms_to_xyz()
+# ------------
+function lms_to_xyz(c::LMS{T}, ::XYZ, cat::MatrixCAT) where T
+    @mul3x3lms XYZ{T} cat_invt(cat) c
+end
+
+function lms_to_xyz(c::LMS{T}, wp::XYZ, cat::CAT_BFD_NL) where T
+    wp_lms = @mul3x3xyz LMS{T} BFD wp
+    p = Float64(wp_lms.s ^ -0.0834)
+    @mul3x3(XYZ{T}, BFD_INV, wp_lms.l * c.l,
+                             wp_lms.m * c.m,
+                             wp_lms.s *(c.s^p) )
+end
+
+
+"""
+    adaptation_xyz(c; src_white=WP_DEFAULT, ref_white=WP_DEFAULT, cat=CAT_02())
+    adaptation_xyz(c, src_white, ref_white, cat)
+
+Apply chromatic adaptation to a color.
+This function is useful when a color is converted to another type after the
+adaptation.
+
+Args:
+
+- `c`: An observed color.
+- `src_white`: Adopted or source white corresponding to `c`.
+- `ref_white`: Reference or destination white.
+- `cat`: Chromatic adaptation transform model.
+
+Returns:
+  A adapted color in XYZ.
+"""
+
+function adaptation_xyz(c::Color;
+                        src_white::Color=WP_DEFAULT,
+                        ref_white::Color=WP_DEFAULT,
+                        cat::AbstractCAT=CAT_02())
+    # this is the wrapper method
+
+    adaptation_xyz(c, src_white, ref_white, cat)
+end
+
+function adaptation_xyz(c::Color, src_white::Color, ref_white::Color, cat::MatrixCAT)
+    c_lms = convert(LMS, c, src_white, cat=cat)
+    src_wp = convert(LMS, src_white, src_white, cat=cat)
+    ref_wp = convert(LMS, ref_white, src_white, cat=cat)
+    ans = LMS(c_lms.l * ref_wp.l / src_wp.l,
+              c_lms.m * ref_wp.m / src_wp.m,
+              c_lms.s * ref_wp.s / src_wp.s)
+    lms_to_xyz(ans, ref_white, cat)
+end
+
+function adaptation_xyz(c::Color, src_white::Color, ref_white::Color, cat::NonLinearCAT)
+    ans = convert(LMS, c, src_white, cat=cat)
+    lms_to_xyz(ans, ref_white, cat)
+end
+
+"""
+    adaptation(c; src_white, ref_white, cat::AbstractCAT=CAT02())
+
+Apply chromatic adaptation to a color.
+
+Args:
+
+- `c`: An observed color.
+- `src_white`: Adopted or source white corresponding to `c`.
+- `ref_white`: Reference or destination white.
+- `cat`: Chromatic adaptation transform model.
+
+Returns:
+  A adapted color.
+"""
+function adaptation(c::T;
+                    src_white::Color=WP_DEFAULT,
+                    ref_white::Color=WP_DEFAULT,
+                    cat::AbstractCAT=CAT02()) where T <: Color
```

Comment:
Terrible typo!
Since the const `CAT02` is not exported, the same mistake will not be repeated in user codes.
```suggestion
                    cat::AbstractCAT=CAT_02()) where T <: Color
```